### PR TITLE
 Implement HTML recursive chunker

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose

Fixes:https://github.com/ballerina-platform/ballerina-library/issues/8170

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

### CODE REVIEW

File: ballerina/chunkers.bal [hunk 1]
```diff
@@ -91,6 +91,39 @@ public isolated class MarkdownChunker {
     }
 }
 
+# Represents a HTML document chunker.
+# Provides functionality to recursively chunk a HTML document using a configurable strategy.
+#
+# The chunking process begins with the specified strategy and recursively falls back to
+# finer-grained strategies if the content exceeds the configured `maxChunkSize`. Overlapping content
+# between chunks can be controlled using `maxOverlapSize`.
+public isolated class HTMLChunker {
```
Comment (MohamedSabthar) [inline ballerina/chunkers.bal:100]:
```suggestion
public isolated class HtmlChunker {
```
Shall we check other places?

File: native/src/test/resources/html-chunker-test/input/very-large.html [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    
+    <h3>Concluding Thoughts and Summary</h3>
+    
+    <p>Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et molestiae non recusandae. Itaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat. Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt.</p>
+    
+    <p>This concludes the very large HTML document designed for stress testing the chunker implementation. The content should be substantial enough to create multiple chunks and test various edge cases in the chunking algorithm.</p>
+</body>
+</html>
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply native/src/test/resources/html-chunker-test/input/very-large.html:57]: Let's fix other resources

File: native/src/test/java/io/ballerina/stdlib/ai/TestUtil.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package io.ballerina.stdlib.ai;
```
Comment (MohamedSabthar) [inline native/src/test/java/io/ballerina/stdlib/ai/TestUtil.java:1]: Let's add the license header

File: ballerina/document-types.bal [hunk 4]
```diff
@@ -41,6 +41,18 @@ public type Metadata record {|
...
+    # Header if chunk belongs to h2 header
+    string h2?;
+    # Header if chunk belongs to h3 header
+    string h3?;
+    # Header if chunk belongs to h4 header
+    string h4?;
+    # Header if chunk belongs to h5 header
+    string h5?;
+    # Header if chunk belongs to h6 header
+    string h6?;
```
Comment (MohamedSabthar) [inline ballerina/document-types.bal:55]: Instead of this can't we have a int field like heading level or something ?

File: ballerina/rag.bal [hunk 5]
```diff
@@ -151,10 +151,18 @@ isolated function guessChunker(Document|Chunk doc) returns Chunker {
...
     }
+    if mimeType == "text/html" {
+        return new HtmlChunker();
+    }
     // Fallback to file name
     string? fileName = doc.metadata?.fileName;
-    if fileName is string && fileName.endsWith(".md") {
-        return new MarkdownChunker();
+    if fileName is string {
+        if fileName.endsWith(".md") {
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply ballerina/rag.bal:160]: My understanding is we support any document type for html and markdown we have specific ones and fallback to generic recursive chunker

File: native/src/main/java/io/ballerina/stdlib/ai/Chunkers.java [hunk 6]
```diff
@@ -90,6 +90,26 @@ public static Object chunkMarkdownDocument(BMap<BString, Object> document, int c
         }
     }
 
+    public static Object chunkHtmlDocument(BMap<BString, Object> document, int chunkSize, int maxOverlapSize,
```
[the remark this answers is not in the document]
Comment (MohamedSabthar) [reply native/src/main/java/io/ballerina/stdlib/ai/Chunkers.java:93]: @heshanpadmasiri was this decision made for performance reasons? We have used an existing java implementation earlier for generic recursive chunker and simply wrapped it for our use-case. But yes we can move the new Html and Markdown chunker implementation to ballerina. @MaryamZi may be we could do this as a separate task.

[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply native/src/main/java/io/ballerina/stdlib/ai/Chunkers.java:93]: The default fallback chunker was in Java because it is a wrapper around langchain4j's DocumentSplitter.  For markdown and HTML, we ended up writing our own (We started considering that we may need to wrap some parser at least). Still, I decided to stick with Java because it is compute-bound (and potentially needs a lot of memory for large files), yet for the time being, I thought spending a lot of cycles trying to polish the algorithm was not worth it.  So more efficient language will probably give us more mileage out of my inefficient algorithm.

**Review by MohamedSabthar — APPROVED** [review]
